### ISSUE TITLE
RandomJigsaw: ensure_perm=True does not exclude the image-preserving permutation (it rejects arange(n), which transposes the patch grid)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4703]
## Summary

`RandomJigsaw(ensure_perm=True)` (the default) is documented as ensuring "the nonidentical patch permutation generation against the original one", i.e. that the output is never the input. It does not: for any grid with both sides greater than 1 the image comes back **unchanged** at the same rate with `ensure_perm=True` as with `ensure_perm=False`.

```python
import torch, kornia.augmentation as K
x = torch.arange(4.).view(2, 2).repeat_interleave(2, 0).repeat_interleave(2, 1)[None, None]   # 2x2 patches of 2x2 px
for ep in (True, False):
    aug = K.RandomJigsaw(grid=(2, 2), p=1.0, ensure_perm=ep); same = 0
    for s in range(2000):
        torch.manual_seed(s); same += int(torch.equal(aug(x), x))
    print(ep, same)
# True  88     <- 4.4 % of calls are a no-op, always with permutation [0, 2, 1, 3]
# False 86
```

Expected rate for a uniform permutation of 4 patches is `1/24 = 4.2 %`; with one permutation excluded, `1/23 = 4.3 %`. So `ensure_perm` excludes a permutation, just not the one that reproduces the image. On a `(2, 3)` grid the no-op rate is 3/2000 either way; only `1 x N` / `N x 1` grids behave as documented (0/2000 vs 1012/2000 for `(1, 2)`).

Measured on `main` @ ff204fb9c, torch 2.14.0, CPU float32.

## Cause

`kornia/augmentation/random_generator/utils.py::randperm` rejects `perm == arange(n)`. `RandomJigsaw.apply_transform` does not use `arange(n)` as the identity, though: an entry's **position** picks the destination cell in column-major order while its **value** indexes the source patch in row-major order (`unfold` yields row-major patches; the final `reshape(-1, b, grid[1], h, piece_w).permute(...)` lays them out column by column). The two orders differ, so

```python
aug = K.RandomJigsaw(grid=(2, 2), p=1.0)
ones = torch.ones(1)
aug(x, params={"batch_prob": ones, "permutation": torch.tensor([[0, 1, 2, 3]])})[0, 0, ::2, ::2]   # [[0, 2], [1, 3]]  transposed
aug(x, params={"batch_prob": ones, "permutation": torch.tensor([[0, 2, 1, 3]])})[0, 0, ::2, ::2]   # [[0, 1], [2, 3]]  the input
```

The image-preserving permutation is `arange(n).view(rows, cols).T.flatten()`, which `ensure_perm` never excludes. The index-identity it does exclude is a genuine shuffle (a transpose of the patch grid on a square grid).

## Expected

Either make the layout consistent, so that `arange(n)` is the identity (and `ensure_perm` is then correct as written), or have `ensure_perm` reject `arange(n).view(rows, cols).T.flatten()`. The first also removes the surprising "identity permutation transposes the image" behaviour; it changes which output a recorded `permutation` replays, so it needs a changelog entry.

Found while reviewing #4634 (batch 6d of #4407), which documents the position/value layout and now cites this issue.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
